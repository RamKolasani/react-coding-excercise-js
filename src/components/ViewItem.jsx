import React, { Component } from 'react';
const AVAILABLE = "Available";
const OUT_OF_STOCK = "Out of stock";

export class ViewItem extends Component {
    constructor(props){
        super(props);
        this.state={ 
            item:{},
            editPriceEnabled:false,
            editAvailabilityEnabled:false,
            availability:"",
            price:0
        }
    }

    componentDidMount(){
        let menuItem = this.props.location.state;
        this.setState({
            item:menuItem,
            availability:menuItem.availability === OUT_OF_STOCK ? false : true,
            price:menuItem.price
            
        });
    }

    handleEditAvailability=()=>{
        this.setState({editAvailabilityEnabled:true});
    }

    handleEditPrice=()=>{
        this.setState({editPriceEnabled:true});
    }

    onChangeAvailablitity =  (e) =>{
        let isAvailable = e.target.value;
        this.setState({availability:isAvailable})
    }

    onChangePrice = (e) =>{
       let safeSearchTypeValue = e.target.value;
       this.setState({price:parseInt(safeSearchTypeValue)});
    }

    handleSave = () =>{
        let isAvailable = this.state.availability === OUT_OF_STOCK ? false : true;
        let newPrice = this.state.price;
        this.setState({ item: { ...this.state.item, price:newPrice, available:isAvailable,}, editAvailabilityEnabled:false, editPriceEnabled:false});
    }

    handleCancel = () =>{
        this.setState({editPriceEnabled:false,
            editAvailabilityEnabled:false})
    }
    
    render() {
        let  item = this.state.item;
        let imageAltText = item.name+" image";
        return (
            <div className="view-item">
                <div id="nav-bar-button" onClick={() =>{this.props.history.push('/')}}> &lt; Back to Dashboard</div>
                <div id="content">
                    <img src={item.imageUrl} alt={imageAltText} />
                    <div id="view-item-details">
                        <p><strong>Item Name: </strong>{item.name}</p>
                        <p><strong>Description: </strong>{item.description}</p>
                        <p><strong>Item Category: </strong>{item.category}</p>
                        <p><strong>Item Id: </strong>{item.itemId}</p>

                        {this.state.editAvailabilityEnabled 
                        ? <span><p><strong>Availability: </strong></p>
                            <select className="availability-selector" name="availability" id="availability" onChange={e=>{this.onChangeAvailablitity(e)}}  >
                            <option value="none" selected disabled hidden> 
                                Select an Option 
                            </option> 
                                <option value="Available" >Available</option>
                                <option value="Out of stock" >Out of Stock</option>
                            </select></span>
                        : <p><strong>Availability: </strong><span data-test="available" >{item.available ? AVAILABLE: OUT_OF_STOCK}</span> <span title="Edit Availability" className="edit-pen" onClick={this.handleEditAvailability}>&#9998;</span></p>
                        }

                        {this.state.editPriceEnabled 
                        ? <span>
                            <p><strong>Item Price: </strong></p>
                            <input className="price-text" type="text" name="price" value={this.state.price} onChange={e => this.onChangePrice(e)}></input>
                        </span>
                        : <p><strong>Item Price: </strong> &#8377;<span data-test="price" >{item.price}</span> <span title="Edit Price" className="edit-pen" onClick={this.handleEditPrice}>&#9998;</span></p>
                        }
                        
                        <p><strong>Tax: </strong> &#8377;{item.tax}</p>
                        {this.state.editAvailabilityEnabled || this.state.editPriceEnabled 
                        ? <div>
                            <button id="save-button" onClick={this.handleSave}>Save</button>
                            <button id="cancel-button" onClick={this.handleCancel}>Cancel</button>
                          </div>:""
                        }
                        
                    </div>
                </div>

            </div>
        )
    }
}

export default ViewItem;
