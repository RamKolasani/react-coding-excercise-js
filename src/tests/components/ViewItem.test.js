import React from 'react';
import { shallow } from 'enzyme';
import ViewItem from '../../components/ViewItem';



describe('testing ViewItem component',()=> {
    let wrapper;
    beforeEach(()=>{
         wrapper = shallow(<ViewItem location={location}/>);
    })

    test('should test ViewItem component',()=>{
        expect(wrapper).toMatchSnapshot();
    });

    test('should render view item page',()=>{
        expect(wrapper.find('div.view-item')).toHaveLength(1);
    });

    test('should render Back to Dashboard button',()=>{
        expect(wrapper.find('div#nav-bar-button')).toHaveLength(1);
    })

    test('should render view page item image',()=>{
        expect(wrapper.find('img')).toHaveLength(1);
    })

    test('should render all item labels',()=>{
        expect(wrapper.find('p')).toHaveLength(7);
        expect(wrapper.find('strong')).toHaveLength(7);
    })

    test('should make Availability into dropdown',()=>{
        expect(wrapper.find('select.availability-selector')).toHaveLength(0);
        expect(wrapper.find('span.edit-pen')).toHaveLength(2);
        wrapper.find('span.edit-pen').at(0).simulate('click');
        expect(wrapper.find('span.edit-pen')).toHaveLength(1);
        expect(wrapper.find('select.availability-selector')).toHaveLength(1);
    })

    test('should make Price into input text',()=>{
        expect(wrapper.find('input.price-text')).toHaveLength(0);
        expect(wrapper.find('span.edit-pen')).toHaveLength(2);
        wrapper.find('span.edit-pen').at(1).simulate('click');
        expect(wrapper.find('span.edit-pen')).toHaveLength(1);
        expect(wrapper.find('input.price-text')).toHaveLength(1);
    })

    test('test cancel functionality',()=>{
        wrapper.find('span.edit-pen').at(0).simulate('click');
        expect(wrapper.find('select.availability-selector')).toHaveLength(1);
        expect(wrapper.find('button')).toHaveLength(2);
        wrapper.find('button').at(1).simulate('click');
        expect(wrapper.find('select.availability-selector')).toHaveLength(0);
    })

    test('test save functionality',()=>{
        wrapper.find('span.edit-pen').at(1).simulate('click');
        expect(wrapper.find('input.price-text')).toHaveLength(1);
        expect(wrapper.find('button')).toHaveLength(2);
        wrapper.find('button').at(0).simulate('click');
        expect(wrapper.find('button')).toHaveLength(0);
    })

    
    test('should edit Price',()=>{
        wrapper.find('span.edit-pen').at(1).simulate('click');
        expect(wrapper.find('input.price-text')).toHaveLength(1);
        wrapper.find('input.price-text').simulate("change", { target: { value: 4000 }});
        wrapper.find('button').at(0).simulate('click');
       expect( wrapper.find('[data-test="price"]').text()).toBe("4000");
    })

    test('should change Availability',()=>{
        wrapper.find('span.edit-pen').at(0).simulate('click');
        expect(wrapper.find('select.availability-selector')).toHaveLength(1);
        wrapper.find('select').simulate('change', {target : { value : 'Out of stock'}});
        wrapper.find('button').at(0).simulate('click');
        expect( wrapper.find('[data-test="available"]').text()).toBe("Out of stock");
    })

    test('test back to dashboard button',()=>{
        const historyMock = {push: jest.fn()}
        wrapper = shallow(<ViewItem history={historyMock} location={location}/>);
        wrapper.find('div#nav-bar-button').simulate('click');
        expect(historyMock.push).toHaveBeenCalled();
})
});





const location ={"state":{
    "name": "Coffee",
    "category": "Beverages",
    "itemId": 16306,
    "imageUrl": "https://presto-resources.s3.amazonaws.com/android__menueditor_images__/images/dynamic/Jw5kV6JUEeqSXYyFkDJJSQ.jpg",
    "price": 200,
    "tax": 8.25,
    "available": true,
    "description": "",
    "modifierGroups": {},
    "timePeriods": [
      {
        "price": null,
        "description": null,
        "daysOfWeek": [],
        "period": [],
        "alwaysAvailable": true
      }
    ]
  }};