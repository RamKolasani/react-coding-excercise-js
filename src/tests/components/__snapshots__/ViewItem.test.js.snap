// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ViewItem component should test ViewItem component 1`] = `
<div
  className="view-item"
>
  <div
    id="nav-bar-button"
    onClick={[Function]}
  >
     &lt; Back to Dashboard
  </div>
  <div
    id="content"
  >
    <img
      alt="Coffee image"
      src="https://presto-resources.s3.amazonaws.com/android__menueditor_images__/images/dynamic/Jw5kV6JUEeqSXYyFkDJJSQ.jpg"
    />
    <div
      id="view-item-details"
    >
      <p>
        <strong>
          Item Name: 
        </strong>
        Coffee
      </p>
      <p>
        <strong>
          Description: 
        </strong>
      </p>
      <p>
        <strong>
          Item Category: 
        </strong>
        Beverages
      </p>
      <p>
        <strong>
          Item Id: 
        </strong>
        16306
      </p>
      <p>
        <strong>
          Availability: 
        </strong>
        <span
          data-test="available"
        >
          Available
        </span>
         
        <span
          className="edit-pen"
          onClick={[Function]}
          title="Edit Availability"
        >
          ✎
        </span>
      </p>
      <p>
        <strong>
          Item Price: 
        </strong>
         ₹
        <span
          data-test="price"
        >
          200
        </span>
         
        <span
          className="edit-pen"
          onClick={[Function]}
          title="Edit Price"
        >
          ✎
        </span>
      </p>
      <p>
        <strong>
          Tax: 
        </strong>
         ₹
        8.25
      </p>
    </div>
  </div>
</div>
`;
